std/hxcoro/AbstractTask.hx

```diff
@@ -81,15 +142,18 @@ abstract class AbstractTask<T> {
 		}
 	}
 
-	/**
-		Returns `true` if cancellation has been requested. This remains true even if cancellation has completed.
-	**/
-	public function cancellationRequested() {
-		return switch (state) {
+	public function onCancellationRequested(f:()->Void):ICancellationHandle {
```

Comment:
I've been making an effort to avoid storing closures in tasks because I would like to keep serialization a possibility. Plus, in my mind arbitrary callbacks like this diminish the whole "structured" aspect we're going for here.

It might be worthwhile to introduce some kind of `ICloseable` interface for this, though a the same time I'm worried about turning everything too Java.

Reply:
I'm not sure there's a way around storing closures in this case, we want some way to be able to react immediately to cancellation. I'm hoping it doesn't break structured concurrency since it's part of the cancellation token interface and exposes none of the task internals.
I also thought about adding something like `IClosable` or `IHandle` since there are now a few interfaces which just have a `close` function.

Reply:
Wouldn't something like this work?

```haxe
class CloseAndCancel<T> implements ICloseable {
	public final handle:ICloseable;
	public final cont:IContinuation<T>;

	public function new(handle:ICloseable, cont:IContinuation<T>) {
		this.handle = handle;
		this.cont = cont;
	}

	public function close() {
		handle.close();
		cont.resume(null, new CancellationException());
	}
}
```

And then we register this as an `ICloseable` on the task, which can either call its `close()` immediately or store the object for later.

Reply:
I think I'm getting myself in a muddle. In this case would `onCancellationRequested` accept a `IContinuation` instead of a function? In that sample is the `handle` field a handle from a scheduler?

Reply:
Sorry for throwing random pieces of code at you, I should probably just implement it properly instead and see if it works!

```haxe
cancellationHandle = ct.onCancellationRequested(new CloseAndCancel(scheduleHandle, cont))
```

And then:

```haxe
	public function onCancellationRequested(closeable:ICloseable):ICancellationHandle {
		return switch state {
			case Cancelling | Cancelled:
				closeable.close();

				return noOpCancellationHandle;
			case _:
				final container = cancellationCallbacks ??= [];
				final handle    = new CancellationHandle(closeable, container);

				container.push(handle);

				handle;
		}
	}
```

And also have `CancellationHandle` call `func.close()` instead.

Reply:
We don't have to do this right now if it makes your life harder. I certainly like how the closure version looks because it's quite direct and readable, so I'm fine with sticking to it for now. If it gets in the way later, we can still change it.